# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/ipc/BrowserChild.cpp
# Commit: 384b025f56d3
# Full Hash: 384b025f56d3f87e045ecede6b1a16b71c0b5d4f
# Author: David P <daparks@mozilla.com>
# Date: 2024-07-04 16:09:52
# Regressor Bug: 1893119
# File Overlap Count: 1
# Description:
#   Bug 1893119: Part 21 - Separate nsIDragService and nsIDragSession implementations r=gstoll,geckoview-reviewers,rkraesig,win-reviewers,m_kato
#   
#   Split the class inheritance trees for nsIDragService and nsIDragSession.
#   Remember that, before the start of this patch series, the inheritance
#   diagram was:
# ==============================================================================

diff -r beeda5ef3712 -r 384b025f56d3 dom/ipc/BrowserChild.cpp
--- a/dom/ipc/BrowserChild.cpp	Thu Jul 04 07:48:11 2024 +0000
+++ b/dom/ipc/BrowserChild.cpp	Thu Jul 04 07:48:11 2024 +0000
@@ -84,6 +84,7 @@
 #include "nsDeviceContext.h"
 #include "nsDocShell.h"
 #include "nsDocShellLoadState.h"
+#include "nsDragServiceProxy.h"
 #include "nsExceptionHandler.h"
 #include "nsFilePickerProxy.h"
 #include "nsFocusManager.h"
@@ -502,6 +503,7 @@
 NS_IMPL_CYCLE_COLLECTION_TRACE_END
 
 NS_INTERFACE_MAP_BEGIN_CYCLE_COLLECTION(BrowserChild)
+  NS_INTERFACE_MAP_ENTRY_CONCRETE(BrowserChild)
   NS_INTERFACE_MAP_ENTRY(nsIWebBrowserChrome)
   NS_INTERFACE_MAP_ENTRY(nsIInterfaceRequestor)
   NS_INTERFACE_MAP_ENTRY(nsIWindowProvider)
@@ -1919,6 +1921,15 @@
   }
 
   DispatchWidgetEventViaAPZ(localEvent);
+
+  if (aEvent.mMessage == eDragLeave ||
+      aEvent.mMessage == eDragExit) {
+    // If session is still active, remove its target.
+    dragSession = GetDragSession();
+    if (dragSession) {
+      static_cast<nsDragSessionProxy*>(dragSession.get())->SetDragTarget(nullptr);
+    }
+  }
   return IPC_OK();
 }
 
@@ -1971,7 +1982,7 @@
           do_GetService("@mozilla.org/widget/dragservice;1")) {
     nsIWidget* widget = WebWidget();
     dragService->StartDragSession(widget);
-    if (RefPtr<nsIDragSession> session = nsContentUtils::GetDragSession(widget)) {
+    if (RefPtr<nsIDragSession> session = GetDragSession()) {
       session->SetSourceWindowContext(aSourceWindowContext.GetMaybeDiscarded());
       session->SetSourceTopWindowContext(
           aSourceTopWindowContext.GetMaybeDiscarded());
@@ -3889,9 +3900,11 @@
 }
 
 already_AddRefed<nsIDragSession> BrowserChild::GetDragSession() {
-  RefPtr<nsIDragSession> session =
-      nsContentUtils::GetDragSession(mPuppetWidget);
-  return session.forget();
+  return RefPtr(mDragSession).forget();
+}
+
+void BrowserChild::SetDragSession(nsIDragSession* aSession) {
+  mDragSession = aSession;
 }
 
 BrowserChildMessageManager::BrowserChildMessageManager(